source/extensions/filters/http/ext_proc/mutation_utils.cc

```diff
@@ -0,0 +1,50 @@
+#include "extensions/filters/http/ext_proc/mutation_utils.h"
+
+#include "common/http/header_utility.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExternalProcessing {
+
+using Http::LowerCaseString;
+
+void MutationUtils::buildHttpHeaders(const Http::HeaderMap& headers_in,
+                                     envoy::config::core::v3::HeaderMap& headers_out) {
+  headers_in.iterate([&headers_out](const Http::HeaderEntry& e) -> Http::HeaderMap::Iterate {
+    auto* new_header = headers_out.add_headers();
+    new_header->set_key(std::string(e.key().getStringView()));
+    new_header->set_value(std::string(e.value().getStringView()));
+    return Http::HeaderMap::Iterate::Continue;
+  });
+}
+
+void MutationUtils::applyHeaderMutations(
+    const envoy::service::ext_proc::v3alpha::HeaderMutation& mutation, Http::HeaderMap& headers) {
+  for (const auto& remove_header : mutation.remove_headers()) {
+    if (Http::HeaderUtility::isRemovableHeader(remove_header)) {
+      headers.remove(LowerCaseString(remove_header));
+    }
+  }
+
+  for (const auto& sh : mutation.set_headers()) {
```

Comment:
What happens if the control plane tries to set something like `:authorith` or `host`? CC @alyssawilk

Reply:
I think if we're going to allow modification of : headers we should have explicit tests.  for example if an endpoint sets 
host: 
instead of 
:authority
I *think* it may affect HTTP/1.1 and not HTTP/2, where I believe :authority would affect both. I'd lean towards disallowing host: unless we integration test that concern is incorrect.
If we allow overwriting :authority (which I agree could be useful) we definitely should call it out explicitly in docs, and explicitly test.  Any change to the authority or URL would also trigger another route lookup, right?  That'd be another good thing to test for and again explicitly document. 
Setting :method also is somewhat concerning, as I'm not sure switching GET to HEAD or vice versa couldn't cause problems where some layers might send a body and others might not. I'd be inclined to disallow unless it's needed, but it's fine to allow if we have rigorous testing.
Another thought, I wonder if there should eventually be config for allowing/disallowing x-envoy headers to be set.  I can imagine use cases where one would want to allow customers to do header modifications, but not say redirect the host (to point to another customer's upstreams) and edit the IP (to look trusted).  Again I think if we allow such edits by default (launch and iterate) it should be explicitly called out in the docs so folks are aware of the level of privilege allowed.

Reply:
yeah, I think that'd be reasonable.  I'd just blanket refuse x-envoy since I think it's easier to reject by prefix than figure out which matter and which don't